### PR TITLE
This simplifies the logic of tracking services and checks

### PR DESCRIPTION
**Pull request by diptanu**
And also adds the service ID to check hash.

### CODE REVIEW

File: client/consul.go [hunk 1]
```diff
@@ -146,6 +126,7 @@ func (c *ConsulService) ShutDown() {
 	close(c.shutdownCh)
 }
 
+// Performs calls to sync checks and services periodically
```
Comment (dadgar) [inline client/consul.go]: The Go style for comments is the methodName and then what it does.

> SyncWithConsul is a long lived function that periodically syncs services with Consul.

File: client/consul.go [hunk 2]
```diff
@@ -256,78 +268,72 @@ func (c *ConsulService) registerService(service *structs.Service, task *structs.
...
+	if err := c.client.ServiceRegister(asr); err != nil {
 		c.logger.Printf("[DEBUG] consul: Error while registering service %v with Consul: %v", service.Name, err)
 		mErr.Errors = append(mErr.Errors, err)
 	}
-	checks := c.makeChecks(service, host, port)
-	for _, check := range checks {
-		if err := c.registerCheck(check); err != nil {
+	for _, check := range service.Checks {
+		cr := c.makeCheck(service, check, host, port)
+		if err := c.registerCheck(cr); err != nil {
```
Comment (dadgar) [inline client/consul.go]: Won't there be an error every time if there is no Consul

File: client/consul.go [hunk 3]
```diff
@@ -124,6 +147,7 @@ func (c *ConsulService) Register(task *structs.Task, allocID string) error {
 	return mErr.ErrorOrNil()
 }
 
+// Stops tracking a task for changes to it's services and checks
```
Comment (dadgar) [inline client/consul.go]: Can you fix the comments that don't match the Go style.

Also shouldn't these two mention that it tracks AND registers/deregisters

File: nomad/structs/structs_test.go [hunk 4]
```diff
@@ -399,9 +399,10 @@ func TestDistinctCheckId(t *testing.T) {
 		Interval: 4 * time.Second,
 		Timeout:  3 * time.Second,
 	}
+	serviceId := "123"
 
-	if c1.Hash() == c2.Hash() || c1.Hash() == c3.Hash() || c3.Hash() == c2.Hash() {
-		t.Fatalf("Checks need to be uniq c1: %s, c2: %s, c3: %s", c1.Hash(), c2.Hash(), c3.Hash())
+	if c1.Hash(serviceId) == c2.Hash(serviceId) || c1.Hash(serviceId) == c3.Hash(serviceId) || c3.Hash(serviceId) == c2.Hash(serviceId) {
```
Comment (dadgar) [inline nomad/structs/structs_test.go]: Could pull these out into variables to make it shorter if you want. `hash1`

File: client/consul.go (hunk 2 above)
Comment (diptanu) [inline client/consul.go]: @dadgar yeah that's fine, we are just swallowing the error. PerformSync keeps trying to register with Consul every 5 seconds.

**Pull request merged**